charm/kafka/lib/charms/layer/kafka.py

```diff
@@ -1,57 +1,60 @@
-# Licensed to the Apache Software Foundation (ASF) under one or more
-# contributor license agreements.  See the NOTICE file distributed with
-# this work for additional information regarding copyright ownership.
-# The ASF licenses this file to You under the Apache License, Version 2.0
-# (the "License"); you may not use this file except in compliance with
-# the License.  You may obtain a copy of the License at
-#
-#     http://www.apache.org/licenses/LICENSE-2.0
-#
-# Unless required by applicable law or agreed to in writing, software
-# distributed under the License is distributed on an "AS IS" BASIS,
-# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-# See the License for the specific language governing permissions and
-# limitations under the License.
-
-import ipaddress
 import os
-import re
 import shutil
-import socket
+import yaml
 
-from pathlib import Path
+from base64 import b64encode
+from os import urandom
 
-from charmhelpers.core import hookenv, host
+from charmhelpers.core import hookenv, host, unitdata
 from charmhelpers.core.templating import render
+from charms.reactive.relations import RelationBase
+from charmhelpers.core.hookenv import log
+from charms.layer import snap
 
+KAFKA_PORT = 9092
+KAFKA_SNAP = 'kafka'
+KAFKA_SERVICE = 'snap.{}.kafka.service'.format(KAFKA_SNAP)
+KAFKA_SNAP_DATA = '/var/snap/{}/common'.format(KAFKA_SNAP)
 
-KAFKA_PORT=9093
-KAFKA_SNAP='kafka'
-KAFKA_SERVICE='snap.{}.kafka.service'.format(KAFKA_SNAP)
-KAFKA_SNAP_DATA='/var/snap/{}/common'.format(KAFKA_SNAP)
 
 class Kafka(object):
+    def __init__(self):
+        '''
+        Initializes a local Kafka object that will contains methods to
+        control the Kafka service snap and also setup required resources.
+        '''
+        if not unitdata.kv().get('kafka.keystore.password'):
```

Comment:
I don't think this belongs in the Kafka constructor. We only need to generate this password once. Also, I don't think it's appropriate to store a secret in unitdata. Passwords and secret keys should not be persisted to disk world readable, which you don't have control over. Best to be explicit and use a root-owner, mode 0600 file for this.